libamqpprox/amqpprox_connectionlimitermanager.cpp

```diff
@@ -0,0 +1,266 @@
+#include <amqpprox_fixedwindowconnectionratelimiter.h>
+#include <amqpprox_logging.h>
+
+#include <memory>
+#include <string>
+
+namespace Bloomberg {
+namespace amqpprox {
+
+ConnectionLimiterManager::ConnectionLimiterManager()
+: d_connectionRateLimitersPerVhost()
+, d_alarmOnlyConnectionRateLimitersPerVhost()
+, d_defaultConnectionRateLimit(0)
+, d_defaultAlarmOnlyConnectionRateLimit(0)
+, d_mutex()
+{
+}
+
+std::shared_ptr<ConnectionLimiterInterface>
+ConnectionLimiterManager::addConnectionRateLimiter(
+    const std::string &vhostName,
+    uint32_t           numberOfConnections)
+{
+    std::lock_guard<std::mutex> lg(d_mutex);
+
+    std::shared_ptr<FixedWindowConnectionRateLimiter> connectionRateLimiter =
+        std::make_shared<FixedWindowConnectionRateLimiter>(
+            numberOfConnections);
+    d_connectionRateLimitersPerVhost[vhostName] = {true,
+                                                   connectionRateLimiter};
+    return connectionRateLimiter;
+}
+
+std::shared_ptr<ConnectionLimiterInterface>
+ConnectionLimiterManager::addAlarmOnlyConnectionRateLimiter(
+    const std::string &vhostName,
+    uint32_t           numberOfConnections)
+{
+    std::lock_guard<std::mutex> lg(d_mutex);
+
+    std::shared_ptr<FixedWindowConnectionRateLimiter>
+        alarmOnlyConnectionRateLimiter =
+            std::make_shared<FixedWindowConnectionRateLimiter>(
+                numberOfConnections);
+    d_alarmOnlyConnectionRateLimitersPerVhost[vhostName] = {
+        true, alarmOnlyConnectionRateLimiter};
+    return alarmOnlyConnectionRateLimiter;
+}
+
+void ConnectionLimiterManager::setDefaultConnectionRateLimit(
+    uint32_t defaultConnectionRateLimit)
+{
+    std::lock_guard<std::mutex> lg(d_mutex);
+
+    d_defaultConnectionRateLimit = defaultConnectionRateLimit;
+    for (auto &limiter : d_connectionRateLimitersPerVhost) {
+        if (!limiter.second.first) {
+            limiter.second.second =
+                std::make_shared<FixedWindowConnectionRateLimiter>(
+                    d_defaultConnectionRateLimit);
+        }
+    }
+}
+
+void ConnectionLimiterManager::setAlarmOnlyDefaultConnectionRateLimit(
+    uint32_t defaultConnectionRateLimit)
+{
+    std::lock_guard<std::mutex> lg(d_mutex);
+
+    d_defaultAlarmOnlyConnectionRateLimit = defaultConnectionRateLimit;
+    // To update new default connection rate limit for all the vhosts,
+    // by removing old already set default connection rate limiters
+    for (auto &limiter : d_alarmOnlyConnectionRateLimitersPerVhost) {
+        if (!limiter.second.first) {
+            limiter.second.second =
+                std::make_shared<FixedWindowConnectionRateLimiter>(
+                    d_defaultAlarmOnlyConnectionRateLimit);
+        }
+    }
+}
+
+void ConnectionLimiterManager::removeConnectionRateLimiter(
+    const std::string &vhostName)
+{
+    std::lock_guard<std::mutex> lg(d_mutex);
+
+    d_connectionRateLimitersPerVhost.erase(vhostName);
```

Comment:
if you check the default first this can go in the else, since its wasteful to erase if you're going to overwrite anyway